File: companion-apps/ios/NavonTests/Fakes/FakeLocationService.swift
import Foundation
@testable import Navon

/// Test double for `LocationService`. Properties are `@Published` and the class
/// conforms to `ObservableObject` so SwiftUI / Combine subscribers see the same
/// change stream they would from the real `CoreLocationService`.
@MainActor
final class FakeLocationService: ObservableObject, LocationService {
    @Published private(set) var currentLocation: CoordinatePoint?
    @Published private(set) var lastKnownLocation: CoordinatePoint?
    @Published private(set) var isLocating: Bool = false
    @Published private(set) var lastError: LocationErrorKind?
    @Published private(set) var currentSpeedMps: Double?

    private(set) var startCount = 0
    private(set) var stopCount = 0

    func start() {
        startCount += 1
        isLocating = true
        lastError = nil
    }

    func stop() {
        stopCount += 1
        isLocating = false
    }

    func setNavigationAccuracy(_ active: Bool) {}

    func emitFix(latitude: Double, longitude: Double, speedMps: Double? = nil) {
        let point = CoordinatePoint(latitude: latitude, longitude: longitude)
        currentLocation = point
        lastKnownLocation = point
        currentSpeedMps = speedMps
        isLocating = false
        lastError = nil
    }

    func emitError(_ kind: LocationErrorKind) {
        lastError = kind
        isLocating = false
    }
}

File: companion-apps/android/app/src/test/java/app/navon/bike/fakes/FakeLocationService.kt
package app.navon.bike.fakes

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import app.navon.bike.domain.CoordinatePoint
import app.navon.bike.domain.LocationErrorKind
import app.navon.bike.domain.LocationService
import app.navon.bike.domain.LocationState

/**
 * Test double for `LocationService`. Tests call `emitFix` / `emitError` to
 * drive the state flow; the real FusedLocationProvider is never touched.
 */
class FakeLocationService : LocationService {
    private val _state = MutableStateFlow(LocationState())
    override val state: StateFlow<LocationState> = _state.asStateFlow()

    override fun hasLocationPermission(): Boolean = true

    var startCount: Int = 0
        private set
    var stopCount: Int = 0
        private set

    override fun start() {
        startCount++
        _state.value = _state.value.copy(isLocating = true, lastError = null)
    }

    override fun stop() {
        stopCount++
        _state.value = _state.value.copy(isLocating = false)
    }

    fun emitFix(latitude: Double, longitude: Double) {
        val point = CoordinatePoint(latitude, longitude)
        _state.value = LocationState(
            currentLocation = point,
            lastKnownLocation = point,
            isLocating = false,
            lastError = null,
        )
    }

    fun emitError(kind: LocationErrorKind) {
        _state.value = _state.value.copy(lastError = kind, isLocating = false)
    }
}
